### ISSUE TITLE
varlociraptor chokes on indels due to Myers pattern_matching

### ISSUE THREAD

**Issue by dlaehnemann** [varlociraptor/varlociraptor#68]
From testing the circleci-provided bioconda package before merging the [varlociraptor recipe PR ](https://github.com/bioconda/bioconda-recipes/pull/15231) over there, I ran into this:

```
[src/model/modes/generic.rs:57] &posterior = GenericPosterior {
    resolutions: [
        100
    ]
}
[src/model/modes/generic.rs:58] &likelihood = GenericLikelihood {
    inner: [
        Normal(
            SampleLikelihoodModel
        )
    ]
}
thread 'main' panicked at 'Pattern too long', /root/.cargo/registry/src/github.com-1ecc6299db9ec823/bio-0.26.1/src/pattern_matching/myers/mod.rs:276:9
stack backtrace:
   0: std::sys::unix::backtrace::tracing::imp::unwind_backtrace
             at src/libstd/sys/unix/backtrace/tracing/gcc_s.rs:39
   1: std::sys_common::backtrace::_print
             at src/libstd/sys_common/backtrace.rs:70
   2: std::panicking::default_hook::{{closure}}
             at src/libstd/sys_common/backtrace.rs:58
             at src/libstd/panicking.rs:200
   3: std::panicking::default_hook
             at src/libstd/panicking.rs:215
   4: std::panicking::rust_panic_with_hook
             at src/libstd/panicking.rs:478
   5: std::panicking::begin_panic
   6: <bio::pattern_matching::myers::Myers<T>>::new
   7: <varlociraptor::model::evidence::reads::IndelEvidence as varlociraptor::model::evidence::reads::AbstractReadEvidence>::prob
   8: varlociraptor::model::sample::Sample::fragment_observation::{{closure}}
   9: varlociraptor::model::sample::Sample::extract_observations
  10: <varlociraptor::call::Caller<A, L, Pr, Po, ModelPayload>>::call
  11: varlociraptor::cli::run
  12: varlociraptor::main
  13: std::rt::lang_start::{{closure}}
  14: std::panicking::try::do_call
             at src/libstd/rt.rs:49
             at src/libstd/panicking.rs:297
  15: __rust_maybe_catch_panic
             at src/libpanic_unwind/lib.rs:87
  16: std::rt::lang_start_internal
             at src/libstd/panicking.rs:276
             at src/libstd/panic.rs:388
             at src/libstd/rt.rs:48
  17: main
  18: __libc_start_main
  19: <unknown>
             at ../sysdeps/x86_64/elf/start.S:103
```

Command used was:
```
RUST_BACKTRACE=1 varlociraptor call variants  --output test_17_E-17_E.bcf --candidates ../genotyped/all.vcf.gz ../data/ref/genome/GCA_000001405.15_GRCh38_no_alt_analysis_set.fna generic --scenario tumour.yaml --bams tumor=../recal/17_E-17_E.bam
```

It is always the first indel that it chokes on and using the `--omit-indels` flag makes it run through. Which makes sense, given it's in the `IndelEvidence` code.

From looking at the `rust-bio` code, the error is happening here:
https://github.com/rust-bio/rust-bio/blob/986e99e968fe3bc9333b58e477321490b84072ca/src/pattern_matching/myers/mod.rs#L272-L278

In the varlociraptor code, it seems to be triggered here:
https://github.com/varlociraptor/varlociraptor/blob/9ef3e6de0c8620e72e29e8e3295a033693307050/src/model/evidence/reads.rs#L272

Something that would make sense as an error source with the pattern length check is that it seems like `varlociraptor` is using the `Myers<u128>`:
https://github.com/varlociraptor/varlociraptor/blob/9ef3e6de0c8620e72e29e8e3295a033693307050/src/model/evidence/reads.rs#L685-L688 

But `rust-bio` until now only implements the `u64` version, if I understand the code correctly:
https://github.com/rust-bio/rust-bio/blob/986e99e968fe3bc9333b58e477321490b84072ca/src/pattern_matching/myers/mod.rs#L255

So I'm guessing it probably requires a bugfix in `rust-bio`, but it's triggered over here, so I figured I'd post it here to start with. 

Also, I'm not sure how to do the implementation. A copy of the `u64` code to drop into place when Type calls for `u128` patterns? Or any way of generalising this from `u64` to both `u64` and `u128`?

**Comment by johanneskoester** [comment]
the u64 in rust-bio is only the default. The problem is rather that varlociraptor chooses a too large window on reads that are longer than 100bp.

**Comment by johanneskoester** [comment]
The solution is to constrain the window length to `record.len() / 2` [here](https://github.com/varlociraptor/varlociraptor/blob/master/src/model/evidence/reads.rs#L216), and rename `self.window` to `self.max_window` in the very same struct. May I ask you to do this today?

**Comment by dlaehnemann** [comment]
I'm on it, but shouldn't `self.max_window` also be capped at 128 bases in case records are longer than 256 bases? Or am I misunderstanding the Myers pattern matching limitations?

**Comment by johanneskoester** [comment]
Yes, that was nonsense what I said. You actually don't need to alter that file. Instead, all needed is to cap the window length at `128 / 2` (because it is applied to the right and the left of the breakpoint. So, this should rather go into cli.rs. And maybe we should not cap it but rather fail with an error if it is set to a larger value.

**Comment by dlaehnemann** [comment]
From playing around with the `--indel-window` parameter in the current circleci build of varlociraptor, I think that we need both. A check that the input parameter is at most 64 and a cap to `read_seq.len() / 2` for when reads are too short (e.g. due to trimming).

**Comment by johanneskoester** [comment]
The latter is already implemented [here](https://github.com/varlociraptor/varlociraptor/blob/master/src/model/evidence/reads.rs#L233) and in the following lines. The calculated window never exceeds the read length.

**Comment by dlaehnemann** [comment]
Hmm, the current circleci build of varlociraptor was still failing in some samples with `--indel-window 64` (as opposed to the default `--indel-window 100`), but not when I decrease it further down to `--indel-window 56`. For this, I have now added a test case with nice new cli tool -- obviously, this test currently fails (`pattern_too_long` is its name).

Also, for now, I have reverted the cap and only left in the input parameter check and error. Interestingly, there is another test that also fails, now:
```
thread 'test26' panicked at '"PROB_SOMATIC_TUMOR >= 4.6" did not return true', tests/lib.rs:181:17
```

Soon to be seen here:
https://travis-ci.org/varlociraptor/varlociraptor/builds/532492578
